### ISSUE TITLE
Compilation fails when using TensorRT 8. 

### ISSUE THREAD

**Issue by feranick** [tensorflow/tensorflow#49150]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Ubuntu 20.04
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if the issue happens on mobile device: N/A
- TensorFlow installed from (source or binary): binary
- TensorFlow version: 2.4.1, 2.5, etc
- Python version: 3.7
- Installed using virtualenv? pip? conda?: no, built from source
- Bazel version (if compiling from source): 3.1 (for TF 2.4.1), 3.7.2 (for TF 2.5.0-rcx)
- GCC/Compiler version (if compiling from source): gcc (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
- CUDA/cuDNN version: Cuda 11.1, cudnn8 (8.0.5.39-1+cuda11.1)
- GPU model and memory: GTX-1080ti
- TensosrRT: 8.0.0-1+cuda11.0

**Describe the problem**
When compiling with support for TensorRT 8 (via libnvinfer8), compilation fails (log is below). 

**Provide the exact sequence of commands / steps that you executed before running into the problem**
When configuring the build, make sure you build with TensorRT support. Build TF as usual.

**Any other info / logs**
Relevant error: 
`C++ compilation of rule '//tensorflow/compiler/tf2tensorrt:tensorrt_stub' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command`

'In file included from bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInfer.h:54,
                 from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:17:
bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInferRuntime.h:2264:51: note: from previous declaration 'nvinfer1::IPluginRegistry* getPluginRegistry() noexcept'
 2264 | extern "C" TENSORRTAPI nvinfer1::IPluginRegistry* getPluginRegistry() noexcept;'

Full log here: 
[gesdm-tf2.5.0rc3-error.txt](https://github.com/tensorflow/tensorflow/files/6469944/gesdm-tf2.5.0rc3-error.txt)

**Comment by feranick** [comment]
Still present for TF 2.5.0

What are the library requirements for building TF2.5.0? Documentation still reference 2.4.0, and that is outdated.

**Comment by ymodak** [comment]
TensorFlow 2.5 pip packages are now built with CUDA11.2 and cuDNN 8.1.0
See https://github.com/tensorflow/tensorflow/releases

**Comment by feranick** [comment]
Compiling without TensorRT support works with cuda-11-2 and cudnn (8.0, 8.1, 8,2). The problem is when support for TensorRT is enabled.

**Comment by ymodak** [comment]
Is the build failing for TF 2.4 as well for you?

**Comment by feranick** [comment]
It fails there too, but I need to do more cross testing. I will report as soon as I am done. Thanks for your patience.

**Comment by feranick** [comment]
Update: it looks like the issue is with tensorRT version 8, regardless of the version of cuda or cudnn, as well as for TF 2.4 or 2.5. When trying to compile with tensorRT 8 (specifically with libnvinfer8, libnvinfer-dev, libnvinfer-plugin8 libnvinfer-plugin-dev), it fails. However, when using tensorRT version 7, it seems to compile just fine. In fact I managed to compile earlier TF 2.5.0rc just fine.

I would have no problem using tensorRT 7. Unfortunately it is no longer available for Ubuntu 20.04. The repos and builds in NVidia repos are only version 8. I was able to test version 7 by manually getting the Ubuntu 18.04 debs and installing them manually on 20.04, which is far less than ideal.

So it seems there may be issues with TF compatibility with tensorRT 8, and NVidia is quite forcing an upgrade that essentially make impossible to build TF with with tensorrt support using Ubuntu 20.04, unless one uses manual installation of libraries.

Happy to run more test.

**Comment by feranick** [comment]
As the last supported version of cuda and libcudnn for tensorrt 7 are 11.1 and 8.0.5, respectively, I am currently trying a couple of builds:
1. cuda-11-2, libcudnn 8.1.1, while retaining tensorrt 7 built for cuda-11-1.
2. Test build using nightly TF 2.6.

**Comment by feranick** [comment]
Cuda-11-2, libcudnn 8.1.1, while retaining tensorrt 7 built for cuda-11-1, can compile just fine.

**Comment by feranick** [comment]
Compiling TF2.6.0-nightly with Cuda-11-2, libcudnn 8.1.1, with TensorRT 8 fails.

**Comment by bixia1** [comment]
This is a known issue, as TensorRT has breaking changes. See [work in progress PR 48917](https://github.com/tensorflow/tensorflow/pull/48917).

**Comment by feranick** [comment]
If it's a known issue it should be reflected in the release notes. It would have saved me quite some time, if I knew...

**Comment by bixia1** [comment]
Do you mean include this in the release note of TensorRT 8, which comes after TF 2.5?

**Comment by feranick** [comment]
Maybe a line specifying that only TensorRT 7 is supported in the 
TensorRT section of the Release notes for TF 2.5...

Alternatively, or in addition, this page is quite outdated (Ubuntu 20.04 
is not even mentioned):

https://www.tensorflow.org/install/gpu

The point is, one could set some very conservative version numbers for 
the GPU libraries needed. However, those are not available in the 
current LTS (and supported) version of Ubuntu, which means that someone 
trying to build TF with it will not be able to. I understand this is 
less of a problem for TF as much as it is an issue in NVidia repos 
dropping support for older versions of TensorRT, but a clarification 
would go a long way.

**Comment by bixia1** [comment]
Interesting point, TensorRT 8 comes after TF 2.5 and we have no ideas that it is not going to work with TF 2.5.

**Comment by feranick** [comment]
Indeed. The issue again is the questionable way NVidia provides packages across major version numbers. To give you a sense, you can actually have both TensorRT 7 and 8 installed through the `libnvinfer8` and `libnvinfer7` and related plugins. However, the dev package (`libnvinfer-dev`) is version agnostic, which means that when NVidia updates TensorRT to version 8, it forces the upgrade also on the dev package. The issue is made worse by the fact that NVidia not only upgraded the packages to 8, but removed all the previous version 7, so there is no way to actively get them through apt. I am using those still available for Ubuntu 18.04, manually installed from packages obtained by decompressing the massive package from NVidia. 

TensorRT has been just released, but Ubuntu20.04 has been around for about a year, which makes this move from NVidia not only questionable, but rather inconsiderate.

**Comment by feranick** [comment]
As to your point about the tensorRT release notes breaking APIs, I can only say that it's fair that is acknowledged. However, you can make that change in such a way that you literally break people systems with no way to recover then in a usable way. As of now, there is no straightforward way to compile tensorflow with tensorRT on Ubuntu 20.04.

**Comment by pseudofunizer** [comment]
> I would have no problem using tensorRT 7. Unfortunately it is no longer available for Ubuntu 20.04. The repos and builds in NVidia repos are only version 8. I was able to test version 7 by manually getting the Ubuntu 18.04 debs and installing them manually on 20.04, which is far less than ideal.
> 
> So it seems there may be issues with TF compatibility with tensorRT 8, and NVidia is quite forcing an upgrade that essentially make impossible to build TF with with tensorrt support using Ubuntu 20.04, unless one uses manual installation of libraries.
> 
> Happy to run more test.

Second this problem on Ubuntu 21.04 (amd64). 
Bazel 3.7.2
Python 3.9.5
CUDA 11.3
cuDNN 8
TensorRT 8 (TensorRT from official Nvidia .deb:
nv-tensorrt-repo-ubuntu2004-cuda11.3-trt8.0.0.3-ea-20210423_1-1_amd64.deb)


```ERROR: /home/luser/tensorflow/tensorflow/compiler/tf2tensorrt/BUILD:43:11: C++ compilation of rule '//tensorflow/compiler/tf2tensorrt:tensorrt_stub' failed (Exit 1): crosstool_wrapper_driver_is_not_gcc failed: error executing command external/local_config_cuda/crosstool/clang/bin/crosstool_wrapper_driver_is_not_gcc -MD -MF bazel-out/k8-opt/bin/tensorflow/compiler/tf2tensorrt/_objs/tensorrt_stub/nvinfer_stub.pic.d ... (remaining 149 argument(s) skipped)
In file included from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:56:
./tensorflow/compiler/tf2tensorrt/stub/NvInfer_5_0.inc:5:7: error: declaration of ‘void* createInferBuilder_INTERNAL(void*, int)’ has a different exception specifier
    5 | void* createInferBuilder_INTERNAL(void* logger, int version) {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:17:
bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInfer.h:8083:30: note: from previous declaration ‘void* createInferBuilder_INTERNAL(void*, int32_t) noexcept’
 8083 | extern "C" TENSORRTAPI void* createInferBuilder_INTERNAL(void* logger, int32_t version) noexcept;
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:56:
./tensorflow/compiler/tf2tensorrt/stub/NvInfer_5_0.inc:12:7: error: declaration of ‘void* createInferRuntime_INTERNAL(void*, int)’ has a different exception specifier
   12 | void* createInferRuntime_INTERNAL(void* logger, int version) {
      |       ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInfer.h:54,
                 from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:17:
bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInferRuntime.h:2253:30: note: from previous declaration ‘void* createInferRuntime_INTERNAL(void*, int32_t) noexcept’
 2253 | extern "C" TENSORRTAPI void* createInferRuntime_INTERNAL(void* logger, int32_t version) noexcept;
      |                              ^~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:56:
./tensorflow/compiler/tf2tensorrt/stub/NvInfer_5_0.inc:33:28: error: declaration of ‘nvinfer1::IPluginRegistry* getPluginRegistry()’ has a different exception specifier
   33 | nvinfer1::IPluginRegistry* getPluginRegistry() {
      |                            ^~~~~~~~~~~~~~~~~
In file included from bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInfer.h:54,
                 from tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc:17:
bazel-out/k8-opt/bin/external/local_config_tensorrt/_virtual_includes/tensorrt_headers/third_party/tensorrt/NvInferRuntime.h:2264:51: note: from previous declaration ‘nvinfer1::IPluginRegistry* getPluginRegistry() noexcept’
 2264 | extern "C" TENSORRTAPI nvinfer1::IPluginRegistry* getPluginRegistry() noexcept;
      |                                                   ^~~~~~~~~~~~~~~~~
Target //tensorflow/tools/pip_package:build_pip_package failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 1411.321s, Critical Path: 125.56s
INFO: 8683 processes: 5111 internal, 3572 local.
FAILED: Build did NOT complete successfully
```

**Comment by chh7411898** [comment]
> ilds in NVidia repos are only version 8. I was able to test version 7 by manually getting the Ubuntu 18.04 debs and installing them manually on 20.04, which is far less than ideal.
> 
> So it seems there may be issues with TF compatibility with tensorRT 8, and NVidia is quite forcing an upgrade that essen

did you solve if?

**Comment by feranick** [comment]
Yes, this is indeed fixed as of TF version 2.7.0On Dec 13, 2021 8:22 PM, hang_code ***@***.***> wrote:

ilds in NVidia repos are only version 8. I was able to test version 7 by manually getting the Ubuntu 18.04 debs and installing them manually on 20.04, which is far less than ideal.
So it seems there may be issues with TF compatibility with tensorRT 8, and NVidia is quite forcing an upgrade that essen

did you solve if?

**Comment by chh7411898** [comment]
thanks  i just modify cmakelist   ,from tensorrt7 to tensorrt 8   fix the problem!

**Comment by aswathselvam** [comment]
Which cmakelist did you modify?

**Comment by bixia1** [comment]
Right, TensorRT has API changes and doesn't work with TF before TF 2.7.

**Comment by SmugglerSMR** [comment]
I faced a similar issue comping TF2.7 with CUDA 11.4, cuDNN 8 and TensorRT 8.2.2 on Python 3.10.1. Unfortunately, I failed to compile due to an unsupported TensorRT version. However, it went almost without issues without it. Except for the bug with Python 3.10.1, which was later fixed with this [commit](https://github.com/tensorflow/tensorflow/commit/14966ee409b89df5b9144a6d1a19359dd0bb1f68?diff=unified).

The good thing is that neither of those issues comes out in version TF2.8.0-rc0. So, if you are interested in adding TensorRT support, you should aim for a version higher than the latest 2.7.0. However, I cannot yet vouch that there will not be any other issue along the way. I had to compile TensortRT from the source for my specific hardware. But according to my short observation, TF2.8 with TensorRT had an execution speed boot.

I used the following code to verify the outcome for my both TF2.7 and TF2.8-RT environments.
```
# %%
# Verefy copiled output
import tensorflow as tf
from tensorflow.compiler.tf2tensorrt._pywrap_py_utils import get_linked_tensorrt_version

from tensorflow.sysconfig import get_build_info

print(f"Compiled TF version {tf.__version__}")

print(f"Linked TensorRT version {get_linked_tensorrt_version()}")

sys_details = get_build_info()
cuda_version = sys_details["cuda_version"]
print(f"Build Cuda version {cuda_version}")

cudnn_version = sys_details["cudnn_version"]  
print(f"Linked cuDNN version {cudnn_version}")

cuda_compute_capabilities = sys_details["cuda_compute_capabilities"]  
print(f"Supported Compute Score {cuda_compute_capabilities}")
# %%
```
Output for TF2.8.0:
```
Compiled TF version 2.8.0-rc0
Linked TensorRT version (8, 2, 2)
Build Cuda version 11.4
Linked cuDNN version 8
Supported Compute Score ['compute_75']
```

**Comment by mylesgoose** [comment]
I tried to build with python 12.3 and Cuda 12.4 and TensoRT version 10 in WSL@ Windows 11pro ubuntu20.04 and the build failed. So i modified theses files. 
Created a new file:NvInfer_10_0.inc

nano ~/tensorflow/tensorflow/compiler/tf2tensorrt/stub/NvInfer_10_0.inc
Content of NvInfer_10_0.inc:

// Auto-generated, do not edit.

extern "C" {

void* createInferBuilder_INTERNAL(void* logger, int version) noexcept {
  using FuncPtr = void * (*)(void *, int);
  static auto func_ptr = LoadSymbol<FuncPtr>("createInferBuilder_INTERNAL");
  if (!func_ptr) LogFatalSymbolNotFound("createInferBuilder_INTERNAL");
  return func_ptr(logger, version);
}

void* createInferRefitter_INTERNAL(void* engine, void* logger,
                                   int version) noexcept {
  using FuncPtr = void* (*)(void*, void*, int);
  static auto func_ptr = LoadSymbol<FuncPtr>("createInferRefitter_INTERNAL");
  if (!func_ptr) LogFatalSymbolNotFound("createInferRefitter_INTERNAL");
  return func_ptr(engine, logger, version);
}

void* createInferRuntime_INTERNAL(void* logger, int version) noexcept {
  using FuncPtr = void * (*)(void *, int);
  static auto func_ptr = LoadSymbol<FuncPtr>("createInferRuntime_INTERNAL");
  if (!func_ptr) LogFatalSymbolNotFound("createInferRuntime_INTERNAL");
  return func_ptr(logger, version);
}

nvinfer1::ILogger* getLogger() noexcept {
  using FuncPtr = nvinfer1::ILogger * (*)();
  static auto func_ptr = LoadSymbol<FuncPtr>("getLogger");
  if (!func_ptr) LogFatalSymbolNotFound("getLogger");
  return func_ptr();
}

int getInferLibVersion() noexcept {
  using FuncPtr = int (*)();
  static auto func_ptr = LoadSymbol<FuncPtr>("getInferLibVersion");
  if (!func_ptr) LogFatalSymbolNotFound("getInferLibVersion");
  return func_ptr();
}

nvinfer1::IPluginRegistry* getPluginRegistry() noexcept {
  using FuncPtr = nvinfer1::IPluginRegistry * (*)();
  static auto func_ptr = LoadSymbol<FuncPtr>("getPluginRegistry");
  if (!func_ptr) LogFatalSymbolNotFound("getPluginRegistry");
  return func_ptr();
}

}  // extern "C"


Modified the files below to add that section to the bottom of each and replace the other part that looks the same i had compiled tensrrt version 10 .:

nano /home/myles1/tensorflow/tensorflow/compiler/tf2tensorrt/stub/nvinfer_stub.cc
nano /home/myles1/tensorflow/tensorflow/compiler/tf2tensorrt/stub/nvinfer_plugin_stub.cc
Content of nvinfer_stub.cc:

#if NV_TENSORRT_MAJOR < 7
#error TensorRT version earlier than 7 is not supported.
#elif NV_TENSORRT_MAJOR == 7
#include "tensorflow/compiler/tf2tensorrt/stub/NvInfer_7_0.inc"
#elif NV_TENSORRT_MAJOR == 8
#include "tensorflow/compiler/tf2tensorrt/stub/NvInfer_8_0.inc"
#elif NV_TENSORRT_MAJOR == 10
#include "tensorflow/compiler/tf2tensorrt/stub/NvInfer_10_0.inc"  // Updated this line
#else
#pragma message("Discovered TensorRT version: " NV_TENSORRT_MAJOR "." NV_TENSORRT_MINOR "." NV_TENSORRT_PATCH " - Using default version 10")
#include "tensorflow/compiler/tf2tensorrt/stub/NvInfer_10_0.inc"  // Default to this line if version is not recognized
#endif
